exercises/roman-numerals/introduction.md

```diff
@@ -0,0 +1,33 @@
+# Description
+
+Today, most of the world use Arabic numbers.
+But if you go back two thousand years, a large proportion of Europe were using Roman numerals instead.
+
+Roman numerals used letters from the Latin alphabet (I, V, X, L, C, D, M) to represent numbers.
+Rather than having digits of 0 to 10, they had numerals representing numbers starting with 1 and 5.
+
+| I | V | X  | L  |  C  |  D  |  M   |
+|---|---|----|----|-----|-----|------|
+| 1 | 5 | 10 | 50 | 100 | 500 | 1000 |
+
+To read Roman numerals you need to know a few basic rules.
+
+- We read from left-to-right.
+- When a letter is repeated multiple times (e.g. `XX`), they get added together: `XX(20) = X(10) + X(10)`
+- The same letter cannot be used more than three times in succession (e.g. `IIII` is invalid - instead you would write `IV`).
+- When a larger letter precedes a smaller letter (e.g. `VI`), letters are added together: `VI(6) = (5) + I(1)`
+- When a smaller letter precedes a larger letter (e.g. `IV`), the first letter is subtracted from the second: `IV(4) = V(5) - I(1)`
+- When subtracting, we cannot have more than one number being subtracted (e.g. `IIV` is invalid) and we can only subtract the nearest two letters (e.g. `IV` and `IX` are valid, but `IL` is not)
+- Use a single letter if possible, (e.g. `VV` is invalid - use `X` instead).
```

Comment:
Note that these rules, as written, mix the writing/converting-to and reading/converting-from perspectives. Some rules are irrelevant when reading, and some rules necessary for writing are missing (e.g. mandatory order by decreasing letter value).

---

There are several useful perspectives, but in one prominent in my head right now

- rules 1 and 3 go together (= + < = &le;),
- rules 2 and 6 go together (minimize number of letters used / length of numeral),
- rule 2+6 provides motivation for subtraction, and
- strictly speaking rule 5 is already covered by rule 4. Well, morally anyway.

Reply:
> (e.g. mandatory order by decreasing letter value).

Is this not the same as "reading left to right", but that needs to be more specified?

----

> There are several useful perspectives, but in one prominent in my head right now

In terms of practically solving this exercise, are there any of these rules that can be left out, or any that need to be added? These were the ones that seemed to be covered during the solutions we looked at in the Deep Dive video.

Reply:
> Is this not the same as "reading left to right", but that needs to be more specified?

The rules as they are now are unclear on XXC. They rule out that it means 80, but seem to allow it to mean 10 + 90 = 100.

Reading order is irrelevant: XLVII can just as well be read as I + I + V + L + -X. But if you specify reading order you can use the words 'before' and 'after' instead of only 'to the left of' and 'to the right of' (which already suffice). Without a specified reading order 'before' and 'after' are meaningless.

( Isn't it weird to speak of reading when the exercise is about writing only? )

One rule that works:

> Ciphers (I V X L C D M) with lesser values are only ever allowed to the right of ciphers with greater values, with a few exceptions:
> - I directly next to V or X,
> - X directly next to L or C, and
> - C directly next to D or M.

This one disallows XXC: the leftmost X is to the left of C, but not directly next to it.

Another one that works:

> Blahblahs\* must be ordered by decreasing value from left to right.

\* Blahblahs could be { I, V, X, L, C, D, M, IV, IX, XL, XC, CD, CM } (i.e. ciphers + subtractions), or what one might call the Roman pseudo-digits: the aforementioned plus the likes of III, LXX, etc. Either works.

---

> In terms of practically solving this exercise, are there any of these rules that can be left out, or any that need to be added? These were the ones that seemed to be covered during the solutions we looked at in the Deep Dive video.

The perspectives I was thinking of are

```
XCVIII = -X + C + V + I + I + I  ciphers
XCVIII = XC + V + I + I + I      no good name found yet
XCVIII = XC + VIII               pseudo-digits, very (too?) easy to work with
```

'Practically solving this exercise' is easiest when given rules in terms of pseudo-digits. These rules are also easy to state, do not include subtraction shenanigans, but do include [this table](https://en.wikipedia.org/wiki/Roman_numerals#Standard_form).

I'll post this now and go through the rules again to properly answer your question.

Reply:
> In terms of practically solving this exercise, are there any of these rules that can be left out, or any that need to be added? These were the ones that seemed to be covered during the solutions we looked at in the Deep Dive video.

Missing rule: only I, X, C are allowed to be subtracted.

Missing rule: something to rule out IXX.

Ambiguity: is IXC allowed?

Reply:
I _think_ this writing explanation is complete:

<details><summary>draft</summary>
To write a Roman numeral we use the following Latin letters, each of which has a value:

| M    | D   | C   | L  | X  | V | I |
| ---- | --- | --- | -- | -- | - | - |
| 1000 | 500 | 100 | 50 | 10 | 5 | 1 |

A Roman numeral is a sequence of these letters, and its value is the sum of the values of its letters.
Example: XVIII has the value 10 + 5 + 1 + 1 + 1 = 18.

Letter order matters.
Letters with greater values must all occur to the left of letters with lesser values.
For example, XMCII is illegal because the value of X (10) is less than the value of C (100).
MCXII on the other hand is legal, and means 1112.

There are six exceptions to the above rule: it is allowed to write

- I immediately to the left of a V or an X,
- X immediately to the left of an L or a C, and
- C immediately to the left of a D or an M.

In these cases the value of this I, X or C is not added, but subtracted.
Example: while CXV means 115, XCV means -10 + 100 + 5 = 95.
Another example: XXC is illegal, because the left X is not directly adjacent to the C.
One more example: CMM is illegal, because the C is not directly adjacent to the second M.

Finally, Roman numerals are always written with the minimum number of letters possible.
That is, instead of VV we write X, and instead of CCCC we write CD.
</details>